backend/server.js

```diff
@@ -62,6 +62,45 @@ hapi.ext({
 sbp('okTurtles.data/set', SERVER_INSTANCE, hapi)
 
 sbp('sbp/selectors/register', {
+  'backend/server/persistState': async function (deserializedHEAD: Object, entry: string) {
+    const contractID = deserializedHEAD.contractID
+    const cheloniaState = sbp('chelonia/private/state')
+    // If the contract has been removed or the height hasn't been updated,
+    // there's nothing to persist
+    if (!cheloniaState.contracts[contractID] || cheloniaState.contracts[contractID].height < deserializedHEAD.head.height) {
+      return
+    }
+    // If the current HEAD is not what we expect, don't save (the state could
+    // have been updated by a later message). This ensures that we save the
+    // latest state and also reduces the number of write operations
+    if (cheloniaState.contracts[contractID].HEAD === deserializedHEAD.hash) {
+      // Extract the parts of the state relevant to this contract
+      const state = {
+        state: cheloniaState[contractID],
+        contractState: cheloniaState.contracts[contractID]
+      }
+      // Save the state under a 'contract partition' key, so that updating a
+      // contract doesn't require saving the entire state.
+      // Although it's not important for the server right now, this will fail to
+      // persist changes the state for other contracts (for example, this happens
+      // when watching keys)
+      await sbp('chelonia/db/set', '_private_cheloniaState_cp_' + contractID, JSON.stringify(state))
+    }
+    // If this is a new contract, we also need to add it to the index, which
+    // is used when starting up the server to know which keys to fetch.
+    // In the future, consider having a multi-level index, since the index can
+    // get pretty large.
+    if (contractID === deserializedHEAD.hash) {
+      // We want to ensure that the index is updated atomically (i.e., if there
+      // are multiple new contracts, all of them should be added), so a queue
+      // is needed for the load & store operation.
+      await sbp('okTurtles.eventQueue/queueEvent', 'update-contract-indices', async () => {
+        const currentIndex = await sbp('chelonia/db/get', '_private_cheloniaState_index')
+        const updatedIndex = `${currentIndex ? `${currentIndex}\x00` : ''}${contractID}`
```

Comment:
What is this `${currentIndex}\x00` weirdness?

(New rule added to Style Guide: [all weirdness must be either eliminated or commented](https://github.com/okTurtles/group-income/blob/master/docs/Style-Guide.md#when-to-write-comments))

Reply:
It doesn't look weird to me, but I can add a comment. It's just adding to the current index (so the result is `${currentIndex}${separator}${contractID}`, with `separator='\x00'`). Since `currentIndex` could be empty or undefined, and we don't want `undefined${separator}${contractID}`, there's a ternary condition.

Reply:
Makes sense. Coming from a C background, seeing `\x00` is especially confusing since that would normally terminate the string 😅

Reply:
I see. I come from a C background too, but `\x00` only terminates a string if you're using the various `str*` functions, which normally you shouldn't use. In any case, I chose `\x00` because it's less magic than other values and also it's probably fair to say that contract IDs can't contain it (in a previous iteration of this change, I was using `,${JSON.stringify(cID)}`), but I figured that this would be simpler.